astropy/coordinates/transformations.py

```diff
@@ -744,8 +748,190 @@ def __call__(self, fromcoord, toframe):
+
+        # We need a velocity (time derivative) and, for now, are strict: the
+        # representation can only contain a velocity differential and no others.
+        if has_velocity and not rad_vel_diff:
+            veldiff = rep.differentials['s'] # already in Cartesian form
+
+            if offset is not None and 's' in offset.differentials:
+                veldiff = veldiff + offset.differentials['s']
+
+            newrep = newrep.with_differentials(veldiff)
+
+        if isinstance(fromcoord.data, UnitSphericalRepresentation):
+            # Special-case this because otherwise the return object will think
+            # it has a valid distance with the default return (a
+            # CartesianRepresentation instance)
+
+            if has_velocity and not unit_vel_diff and not rad_vel_diff:
+                # We have to first represent as the Unit types we converted to,
+                # then put the d_distance information back in to the
+                # differentials and re-represent as their original forms
+                newdiff = newrep.differentials['s']
+                _unit_cls = fromcoord.data.differentials['s']._unit_differential
+                newdiff = newdiff.represent_as(_unit_cls, fromcoord.data)
+
+                kwargs = dict([(comp, getattr(newdiff, comp))
+                               for comp in newdiff.components])
+                kwargs['d_distance'] = fromcoord.data.differentials['s'].d_distance
+                diffs = {'s': fromcoord.data.differentials['s'].__class__(
+                    copy=False, **kwargs)}
+
+            else:
+                diffs = newrep.differentials
+
+            newrep = newrep.represent_as(fromcoord.data.__class__) # drops diffs
+            newrep = newrep.with_differentials(diffs)
+
+        # We pulled the radial differential off of the representation
+        # earlier, so now we need to put it back. But, in order to do that, we
+        # have to turn the representation into a repr that is compatible with
+        # having a RadialDifferential
+        if has_velocity and rad_vel_diff:
+            newrep = newrep.represent_as(fromcoord.data.__class__)
+            newrep = newrep.with_differentials({'s': fromcoord.data.differentials['s']})
 
-class StaticMatrixTransform(CoordinateTransform):
+        return newrep
+
+class AffineTransform(BaseAffineTransform):
+    """
+    A coordinate transformation specified as a function that yields a 3 x 3
+    cartesian transformation matrix and a tuple of displacement vectors.
+
+    Parameters
+    ----------
+    transform_func : callable
+        A callable that has the signature ``transform_func(fromcoord, toframe)``
+        and returns a length-2 tuple that contains: [0] a (3, 3) matrix that
+        converts ``fromcoord`` in a cartesian representation to the new
+        coordinate system, and [1] a (N, 3) set of vectors that contain
+        offsets for differentials (or None).
```

Comment:
Maybe you could add a notes section with a toy example of this functions call/return values (implementation itself not needed although it might be clearer)? I understand it from thinking about it for a while but just reading this description its hard to understand that you mean basically ``M, (origin, velocity_offset) =transform_func(fromcoord, toframe) `` (but with an option for more differentials)